### DISCUSSION TITLE
Middleware Schema

### DISCUSSION THREAD

**Discussion by laurent512** [next-safe-action/next-safe-action#141] [General]
Hello, 

First of all, thank you so much for the great library. It is extremely useful to me. 

When I create a safe action with a middleware, is there a way to add a schema to that middleware ? 
And ensure that it is typesafe ? Somehow if my middleware requires { dataId : string} and the function itself { newData : string} , when calling the safe actions, it should only accept { dataId : string ; newData : string}  ? 

Thank you so much :)

**Comment by TheEdoRan** [comment]
Not sure if I fully understand the issue, but in the upcoming v7 you can pass a [`defineMetadataSchema`](https://next.next-safe-action.dev/docs/safe-action-client/instance-methods#metadata) function to the safe action client and then use the [`metadata`](https://next.next-safe-action.dev/docs/safe-action-client/instance-methods#metadata) method to pass data to middleware and server code functions when defining actions.

**Comment by laurent512** [comment]
Sorry if my explanations were not clear. Let me share with you an example code ; 

```
export const secretAction = createSafeActionClient({
    async middleware(parsedInput, data?: MiddlewareData) {
    
        if (parsedInput.secret != "123456" ) {
            throw new Error(" secret is not correct!");
        } 
        return query;
    },
});
```
With this being defined, I can create a customServerAction like this 

```
export const customServerAction= secretAction (
    customServerActionschema,
    async (parsedData, middlewareReturn) => {
      
        return {
            success:true
        };
    }
);
```

Now If I call that customServerAction somewhere in my code, I would expect some typesafety if its argument are not containing { secret : "..." } because it is required by the middleware. 

I hope this makes it more understable :p 
Maybe the medatata source you pointed does indeed that job

**Reply by TheEdoRan** [reply]
The problem is that middleware functions are parent to safe actions, they're one level above in the execution chain, so it's impossible to guarantee type safety for input from actions in the middleware. If you're really sure that you want a `secret` string as input for every action you're going to define, I think the best way is to manually validate `parsedInput` inside middleware body, using your preferred validation library.

**Reply by laurent512** [reply]
Yeah, let's say I am looking for some typesafety :/ 
Currently, what I do is to remember adding in every action schema I define the required secret string... Ideally I would like to write a schema without it :)

**Reply by TheEdoRan** [reply]
Type safety only works from top to bottom, it's impossible to achieve (inferred) type safety the other way around, this isn't related to next-safe-action, TypeScript works this way. Though, you could define a base secret schema and use it both inside the middleware to validate action input and to define action schemas, by extending it. Hope this helps.
